File: threshold/lib/core/dkg.dart
import 'package:pointycastle/ecc/api.dart';
import 'package:threshold/core/commitment.dart';
import 'package:threshold/core/errors.dart';
import 'package:threshold/core/identifier.dart';
import 'package:threshold/core/share.dart';
import 'package:threshold/core/utils.dart';
import 'package:convert/convert.dart';
import 'dart:typed_data';

class DKGSignature {
  final ECPoint R;
  final BigInt Z;

  DKGSignature(this.R, this.Z);

  factory DKGSignature.fromJson(Map<String, dynamic> json) {
    final R = elemDeserializeCompressed(
      Uint8List.fromList(hex.decode(json['R'])),
    );
    final Z = bytesToBigInt(Uint8List.fromList(hex.decode(json['Z'])));
    return DKGSignature(R, Z);
  }

  Map<String, dynamic> toJson() => {
    'R': hex.encode(elemSerializeCompressed(R)),
    'Z': hex.encode(bigIntToBytes(Z)),
  };
}

class Round1Package {
  final VerifiableSecretSharingCommitment commitment;
  final DKGSignature proofOfKnowledge;

  Round1Package(this.commitment, this.proofOfKnowledge);

  factory Round1Package.fromJson(Map<String, dynamic> json) {
    return Round1Package(
      VerifiableSecretSharingCommitment.fromJson(json['commitment']),
      DKGSignature.fromJson(json['proofOfKnowledge']),
    );
  }

  Map<String, dynamic> toJson() => {
    'commitment': commitment.toJson(),
    'proofOfKnowledge': proofOfKnowledge.toJson(),
  };
}

class Challenge {
  final BigInt C;
  Challenge(this.C);
}

class Round1SecretPackage {
  final Identifier identifier;
  final List<BigInt> coefficients;
  final VerifiableSecretSharingCommitment commitment;
  final int minSigners;
  final int maxSigners;

  Round1SecretPackage(
    this.identifier,
    this.coefficients,
    this.commitment,
    this.minSigners,
    this.maxSigners,
  );
}

class Round2Package {
  final SecretShare secretShare;

  Round2Package(this.secretShare);

  factory Round2Package.fromJson(Map<String, dynamic> json) {
    return Round2Package(
      bytesToBigInt(Uint8List.fromList(hex.decode(json['secretShare']))),
    );
  }

  Map<String, dynamic> toJson() => {
    'secretShare': hex.encode(bigIntToBytes(secretShare)),
  };
}

class Round2SecretPackage {
  final Identifier identifier;
  final VerifiableSecretSharingCommitment commitment;
  final BigInt secretShare;
  final int minSigners;
  final int maxSigners;

  Round2SecretPackage(
    this.identifier,
    this.commitment,
    this.secretShare,
    this.minSigners,
    this.maxSigners,
  );
}

class KeyPackage {
  final Identifier identifier;
  final SecretShare secretShare;
  final VerifyingShare verifyingShare;
  final VerifyingKey verifyingKey;
  final int minSigners;

  KeyPackage(
    this.identifier,
    this.secretShare,
    this.verifyingShare,
    this.verifyingKey,
    this.minSigners,
  );

  factory KeyPackage.fromJson(Map<String, dynamic> json) {
    return KeyPackage(
      Identifier.deserialize(
        Uint8List.fromList(hex.decode(json['identifier'])),
      ),
      bytesToBigInt(Uint8List.fromList(hex.decode(json['secretShare']))),
      elemDeserializeCompressed(
        Uint8List.fromList(hex.decode(json['verifyingShare'])),
      ),
      VerifyingKey(
        E: elemDeserializeCompressed(
          Uint8List.fromList(hex.decode(json['verifyingKey'])),
        ),
      ),
      json['minSigners'],
    );
  }

  Map<String, dynamic> toJson() => {
    'identifier': hex.encode(identifier.serialize()),
    'secretShare': hex.encode(bigIntToBytes(secretShare)),
    'verifyingShare': hex.encode(elemSerializeCompressed(verifyingShare)),
    'verifyingKey': hex.encode(elemSerializeCompressed(verifyingKey.E)),
    'minSigners': minSigners,
  };

  bool get hasEvenY {
    return verifyingKey.E.y!.toBigInteger()!.isEven;
  }

  KeyPackage intoEvenY({bool? isEven}) {
    final currentIsEven = isEven ?? hasEvenY;
    if (!currentIsEven) {
      // Negate all components
      final n = secp256k1Curve.n;
      final negMultiplier = n - BigInt.one;

      final newSecretShare = (n - secretShare) % n;
      // Negate points by multiplying by (n-1) since .negate() is missing
      final newVerifyingShare = (verifyingShare * negMultiplier)!;
      final newVerifyingKeyPoint = (verifyingKey.E * negMultiplier)!;

      return KeyPackage(
        identifier,
        newSecretShare,
        newVerifyingShare,
        VerifyingKey(E: newVerifyingKeyPoint),
        minSigners,
      );
    }
    return this;
  }
}

class PublicKeyPackage {
  final Map<Identifier, VerifyingShare> verifyingShares;
  final VerifyingKey verifyingKey;

  PublicKeyPackage(this.verifyingShares, this.verifyingKey);

  factory PublicKeyPackage.fromJson(Map<String, dynamic> json) {
    final verifyingShares = (json['verifyingShares'] as Map<String, dynamic>)
        .map((key, value) {
          final id = Identifier.deserialize(
            Uint8List.fromList(hex.decode(key)),
          );
          final share = elemDeserializeCompressed(
            Uint8List.fromList(hex.decode(value)),
          );
          return MapEntry(id, share);
        });
    final verifyingKey = VerifyingKey(
      E: elemDeserializeCompressed(
        Uint8List.fromList(hex.decode(json['verifyingKey'])),
      ),
    );
    return PublicKeyPackage(verifyingShares, verifyingKey);
  }

  Map<String, dynamic> toJson() {
    final verifyingShares = this.verifyingShares.map(
      (key, value) => MapEntry(
        hex.encode(key.serialize()),
        hex.encode(elemSerializeCompressed(value)),
      ),
    );
    return {
      'verifyingShares': verifyingShares,
      'verifyingKey': hex.encode(elemSerializeCompressed(verifyingKey.E)),
    };
  }

  bool get hasEvenY {
    return verifyingKey.E.y!.toBigInteger()!.isEven;
  }

  PublicKeyPackage intoEvenY({bool? isEven}) {
    final currentIsEven = isEven ?? hasEvenY;
    if (!currentIsEven) {
      final n = secp256k1Curve.n;
      final negMultiplier = n - BigInt.one;

      final newVerifyingKeyPoint = (verifyingKey.E * negMultiplier)!;
      final newVerifyingShares = verifyingShares.map((id, share) {
        final newShare = (share * negMultiplier)!;
        return MapEntry(id, newShare);
      });

      return PublicKeyPackage(
        newVerifyingShares,
        VerifyingKey(E: newVerifyingKeyPoint),
      );
    }
    return this;
  }
}

class SecretKey {
  final BigInt scalar;
  SecretKey(this.scalar);
}

SecretShare secretShareFromCoefficients(List<BigInt> coeffs, Identifier peer) {
  return evaluatePolynomial(peer, coeffs);
}

(Round1SecretPackage, Round1Package) dkgPart1(
  Identifier identifier,
  int maxSigners,
  int minSigners,
  SecretKey secretKey,
  List<BigInt> coffiecients,
) {
  validateNumOfSigners(minSigners, maxSigners);

  final (coeffs, commitment) = generateSecretPolynomial(
    secretKey.scalar,
    maxSigners,
    minSigners,
    coffiecients,
  );

  final verifyingCommit = VerifiableSecretSharingCommitment(commitment);
  final verifyingKey = verifyingCommit.toVerifyingKey();

  final sig = computeProofOfKnowledge(identifier, coeffs, verifyingKey);

  final secretPkg = Round1SecretPackage(
    identifier,
    coeffs,
    VerifiableSecretSharingCommitment(commitment),
    minSigners,
    maxSigners,
  );
  final pubPkg = Round1Package(
    VerifiableSecretSharingCommitment(commitment),
    sig,
  );
  return (secretPkg, pubPkg);
}

(Round2SecretPackage, Map<Identifier, Round2Package>) dkgPart2(
  Round1SecretPackage secretPkg,
  Map<Identifier, Round1Package> round1Pkgs,
) {
  if (round1Pkgs.length != secretPkg.maxSigners - 1) {
    throw IncorrectNumberOfPackagesException("incorrect number of packages");
  }
  for (final p in round1Pkgs.values) {
    if (p.commitment.coeffs.length != secretPkg.minSigners) {
      throw IncorrectNumberOfCommitmentsException(
        "incorrect number of commitments",
      );
    }
  }

  final out = <Identifier, Round2Package>{};
  for (final entry in round1Pkgs.entries) {
    final senderID = entry.key;
    final pkg = entry.value;

    final verifyingKey = pkg.commitment.toVerifyingKey();
    verifyProofOfKnowledge(senderID, verifyingKey, pkg.proofOfKnowledge);

    final share = secretShareFromCoefficients(secretPkg.coefficients, senderID);
    out[senderID] = Round2Package(share);
  }

  final fii = evaluatePolynomial(secretPkg.identifier, secretPkg.coefficients);

  return (
    Round2SecretPackage(
      secretPkg.identifier,
      secretPkg.commitment,
      fii,
      secretPkg.minSigners,
      secretPkg.maxSigners,
    ),
    out,
  );
}

(KeyPackage, PublicKeyPackage) dkgPart3(
  Round1SecretPackage r1Secret,
  Round2SecretPackage r2Secret,
  Map<Identifier, Round1Package> round1Pkgs,
  Map<Identifier, Round2Package> round2Pkgs,
) {
  if (round1Pkgs.length != r2Secret.maxSigners - 1) {
    throw IncorrectNumberOfPackagesException("incorrect number of packages");
  }
  if (round1Pkgs.length != round2Pkgs.length) {
    throw IncorrectNumberOfPackagesException("incorrect number of packages");
  }
  for (final id in round1Pkgs.keys) {
    if (!round2Pkgs.containsKey(id)) {
      throw IncorrectPackageException("incorrect package mapping");
    }
  }

  var si = modNZero();

  for (final entry in round2Pkgs.entries) {
    final senderID = entry.key;
    final pkg2 = entry.value;

    final r1 = round1Pkgs[senderID]!;
    final temp = ThresholdShare(
      r2Secret.identifier,
      pkg2.secretShare,
      elemBaseMul(pkg2.secretShare),
      r1.commitment,
    );
    temp.verify();
    si = (si + pkg2.secretShare);
  }

  si = (si + r2Secret.secretShare);
  final secretShare = si;

  final verifyingShare = elemBaseMul(secretShare);

  final commitMap = <Identifier, VerifiableSecretSharingCommitment>{};
  for (final entry in round1Pkgs.entries) {
    commitMap[entry.key] = entry.value.commitment;
  }
  commitMap[r2Secret.identifier] = r2Secret.commitment;

  final publicKeyPackage = pkpFromDkgCommitments(commitMap);

  final keyPackage = KeyPackage(
    r2Secret.identifier,
    secretShare,
    verifyingShare,
    publicKeyPackage.verifyingKey,
    r2Secret.minSigners,
  );

  return (keyPackage.intoEvenY(), publicKeyPackage.intoEvenY());
}

PublicKeyPackage pkpFromDkgCommitments(
  Map<Identifier, VerifiableSecretSharingCommitment> commits,
) {
  final ids = commits.keys.toList();
  final list = commits.values.toList();

  final group = sumCommitments(list);
  ids.sort((a, b) => a.s.compareTo(b.s));
  return pkpFromCommitment(ids, group);
}

PublicKeyPackage pkpFromCommitment(
  List<Identifier> ids,
  VerifiableSecretSharingCommitment commit,
) {
  final vmap = <Identifier, VerifyingShare>{};
  for (final id in ids) {
    vmap[id] = commit.getVerifyingShare(id);
  }
  final vk = commit.toVerifyingKey();
  return PublicKeyPackage(vmap, vk);
}

(Round1SecretPackage, Round1Package) dkgRefreshPart1(
  Identifier identifier,
  int maxSigners,
  int minSigners,
) {
  validateNumOfSigners(minSigners, maxSigners);

  final refreshingKey = modNZero();
  final coeffOnly = generateCoefficients(minSigners - 1);

  final (coeffs, commitment) = generateSecretPolynomial(
    refreshingKey,
    maxSigners,
    minSigners,
    coeffOnly,
  );

  if (commitment.isEmpty) {
    throw InvalidCommitVectorException("invalid commit vector");
  }
  final trimmed = commitment.sublist(1);
  final trimCommit = VerifiableSecretSharingCommitment(trimmed);

  final verifyingKey = trimCommit.toVerifyingKey();

  final sig = computeProofOfKnowledge(identifier, coeffs, verifyingKey);

  final sec = Round1SecretPackage(
    identifier,
    coeffs,
    trimCommit,
    minSigners,
    maxSigners,
  );
  final pub = Round1Package(trimCommit, sig);
  return (sec, pub);
}

(Round2SecretPackage, Map<Identifier, Round2Package>) dkgRefreshPart2(
  Round1SecretPackage secretPkg,
  Map<Identifier, Round1Package> round1Pkgs,
) {
  if (round1Pkgs.length != secretPkg.maxSigners - 1) {
    throw IncorrectNumberOfPackagesException("incorrect number of packages");
  }

  final elemIdentity = secp256k1Curve.curve.infinity!;
  final identity = elemIdentity;

  final myCoeffs = <ECPoint>[identity, ...secretPkg.commitment.coeffs];
  secretPkg = Round1SecretPackage(
    secretPkg.identifier,
    secretPkg.coefficients,
    VerifiableSecretSharingCommitment(myCoeffs),
    secretPkg.minSigners,
    secretPkg.maxSigners,
  );

  final out = <Identifier, Round2Package>{};

  for (final entry in round1Pkgs.entries) {
    final senderID = entry.key;
    final r1 = entry.value;

    final peerCoeffs = <ECPoint>[identity, ...r1.commitment.coeffs];

    if (peerCoeffs.length != secretPkg.minSigners) {
      throw IncorrectNumberOfCommitmentsException(
        "incorrect number of commitments",
      );
    }

    final share = secretShareFromCoefficients(secretPkg.coefficients, senderID);
    out[senderID] = Round2Package(share);
  }

  final fii = evaluatePolynomial(secretPkg.identifier, secretPkg.coefficients);

  return (
    Round2SecretPackage(
      secretPkg.identifier,
      secretPkg.commitment,
      fii,
      secretPkg.minSigners,
      secretPkg.maxSigners,
    ),
    out,
  );
}

(KeyPackage, PublicKeyPackage) dkgRefreshPart3(
  Round2SecretPackage r2Secret,
  Map<Identifier, Round1Package> round1Pkgs,
  Map<Identifier, Round2Package> round2Pkgs,
  PublicKeyPackage oldPKP,
  KeyPackage oldKP,
) {
  final newR1 = <Identifier, Round1Package>{};
  final elemIdentity = secp256k1Curve.curve.infinity!;
  final identity = elemIdentity;

  for (final entry in round1Pkgs.entries) {
    final senderID = entry.key;
    final r1 = entry.value;
    final coeffs = <ECPoint>[identity, ...r1.commitment.coeffs];
    newR1[senderID] = Round1Package(
      VerifiableSecretSharingCommitment(coeffs),
      r1.proofOfKnowledge,
    );
  }

  if (newR1.length != r2Secret.maxSigners - 1) {
    throw IncorrectNumberOfPackagesException("incorrect number of packages");
  }
  if (newR1.length != round2Pkgs.length) {
    throw IncorrectNumberOfPackagesException("incorrect number of packages");
  }
  for (final id in newR1.keys) {
    if (!round2Pkgs.containsKey(id)) {
      throw IncorrectPackageException("incorrect package mapping");
    }
  }

  var si = modNZero();

  for (final entry in round2Pkgs.entries) {
    final senderID = entry.key;
    final r2 = entry.value;

    final r1 = newR1[senderID]!;
    final temp = ThresholdShare(
      r2Secret.identifier,
      r2.secretShare,
      elemBaseMul(r2.secretShare),
      r1.commitment,
    );
    temp.verify();
    si = (si + r2.secretShare);
  }

  si = (si + r2Secret.secretShare);

  final oldShare = oldKP.secretShare;
  si = (si + oldShare);

  final newSecretShare = si;
  final newVerifying = elemBaseMul(newSecretShare);

  final commitMap = <Identifier, VerifiableSecretSharingCommitment>{};
  for (final entry in newR1.entries) {
    commitMap[entry.key] = entry.value.commitment;
  }
  commitMap[r2Secret.identifier] = r2Secret.commitment;

  final zeroPKP = pkpFromDkgCommitments(commitMap);

  final newVS = <Identifier, VerifyingShare>{};
  for (final entry in zeroPKP.verifyingShares.entries) {
    final id = entry.key;
    final vsNew = entry.value;
    final vsOld = oldPKP.verifyingShares[id];
    if (vsOld == null) {
      throw UnknownIdentifierException("unknown identifier");
    }
    final sum = elemAdd(vsNew, vsOld);
    newVS[id] = sum;
  }

  final pub = PublicKeyPackage(newVS, oldPKP.verifyingKey);

  final kp = KeyPackage(
    r2Secret.identifier,
    newSecretShare,
    newVerifying,
    pub.verifyingKey,
    r2Secret.minSigners,
  );

  return (kp, pub);
}

File: threshold-rs/src/dkg.rs
use crate::error::Error;
use crate::identifier::Identifier;
use crate::keys::{PublicKeyPackage, VerifyingKey};
use crate::point;
use crate::polynomial;
use crate::scalar::{scalar_from_bytes, scalar_to_bytes};
use crate::vss::{self, VssCommitment};
use alloc::collections::BTreeMap;
use alloc::string::String;
use alloc::vec::Vec;
use k256::elliptic_curve::ops::Reduce;
use k256::{ProjectivePoint, Scalar, U256};
use rand_core::RngCore;
use sha2::{Digest, Sha256};

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

/// Proof-of-knowledge signature used during DKG round 1.
#[derive(Clone, Debug)]
pub struct DkgSignature {
    pub r: ProjectivePoint,
    pub z: Scalar,
}

/// Public package broadcast by each participant after DKG round 1.
#[derive(Clone, Debug)]
pub struct Round1Package {
    pub commitment: VssCommitment,
    pub proof_of_knowledge: DkgSignature,
    pub verifying_key: VerifyingKey,
}

/// Secret state kept by a participant between DKG rounds 1 and 2.
#[derive(Clone, Debug)]
pub struct Round1SecretPackage {
    pub identifier: Identifier,
    pub coefficients: Vec<Scalar>,
    pub commitment: VssCommitment,
    pub min_signers: usize,
    pub max_signers: usize,
}

/// Package sent point-to-point from one participant to another in round 2.
#[derive(Clone, Debug)]
pub struct Round2Package {
    pub secret_share: Scalar,
}

/// Secret state kept by a participant between DKG rounds 2 and 3.
#[derive(Clone, Debug)]
pub struct Round2SecretPackage {
    pub identifier: Identifier,
    pub commitment: VssCommitment,
    pub secret_share: Scalar,
    pub min_signers: usize,
    pub max_signers: usize,
}

// ---------------------------------------------------------------------------
// DKG challenge (matches Dart dkgChallenge in utils.dart:199)
// ---------------------------------------------------------------------------

/// Compute the DKG proof-of-knowledge challenge:
/// c = SHA256(id_bytes(32) || vk_compressed(33) || R_compressed(33)) mod n
fn dkg_challenge(id: &Identifier, vk: &VerifyingKey, r: &ProjectivePoint) -> Scalar {
    let mut hasher = Sha256::new();
    hasher.update(id.serialize());
    hasher.update(vk.serialize());
    hasher.update(point::serialize_compressed(r));
    let hash = hasher.finalize();
    let wide = U256::from_be_slice(&hash);
    <Scalar as Reduce<U256>>::reduce(wide)
}

// ---------------------------------------------------------------------------
// Proof of knowledge (matches Dart computeProofOfKnowledge / verifyProofOfKnowledge)
// ---------------------------------------------------------------------------

/// Generate a random non-zero scalar.
fn random_scalar(rng: &mut impl RngCore) -> Scalar {
    loop {
        let mut bytes = [0u8; 32];
        rng.fill_bytes(&mut bytes);
        let wide = U256::from_be_slice(&bytes);
        let s = <Scalar as Reduce<U256>>::reduce(wide);
        if !bool::from(s.is_zero()) {
            return s;
        }
    }
}

/// Compute a proof of knowledge of the secret polynomial constant term.
///
/// sig.z = a0 * c + k (mod n), sig.r = k * G
pub fn compute_proof_of_knowledge(
    id: &Identifier,
    coefficients: &[Scalar],
    vk: &VerifyingKey,
    rng: &mut impl RngCore,
) -> Result<DkgSignature, Error> {
    if coefficients.is_empty() {
        return Err(Error::InvalidCoefficients);
    }
    let k = random_scalar(rng);
    let r = point::base_mul(&k);

    let c = dkg_challenge(id, vk, &r);
    let a0 = coefficients[0];
    let z = a0 * c + k;

    Ok(DkgSignature { r, z })
}

/// Verify a proof of knowledge.
///
/// Check: R == z*G + (-c)*VK
pub fn verify_proof_of_knowledge(
    id: &Identifier,
    vk: &VerifyingKey,
    sig: &DkgSignature,
) -> Result<(), Error> {
    let c = dkg_challenge(id, vk, &sig.r);
    let z_g = point::base_mul(&sig.z);
    let c_neg = -c;
    let c_neg_vk = point::point_mul(&vk.point, &c_neg);
    let right = point::point_add(&z_g, &c_neg_vk);

    if point::points_equal(&sig.r, &right) {
        Ok(())
    } else {
        Err(Error::InvalidProofOfKnowledge)
    }
}

// ---------------------------------------------------------------------------
// Validation
// ---------------------------------------------------------------------------

fn validate_num_signers(min: usize, max: usize) -> Result<(), Error> {
    if min < 2 {
        return Err(Error::InvalidMinSigners);
    }
    if max < 2 {
        return Err(Error::InvalidMaxSigners);
    }
    if min > max {
        return Err(Error::InvalidMinSigners);
    }
    Ok(())
}

// ---------------------------------------------------------------------------
// DKG Part 1 (matches Dart dkgPart1 in dkg.dart:298)
// ---------------------------------------------------------------------------

/// DKG round 1: generate secret polynomial, commitment, and proof of knowledge.
///
/// - `secret`: the participant's secret (polynomial constant term).
/// - `coefficients`: the remaining polynomial coefficients (length = min_signers - 1).
/// - Returns `(Round1SecretPackage, Round1Package)`.
pub fn dkg_part1(
    max_signers: usize,
    min_signers: usize,
    secret: &Scalar,
    coefficients: &[Scalar],
    rng: &mut impl RngCore,
) -> Result<(Round1SecretPackage, Round1Package), Error> {
    validate_num_signers(min_signers, max_signers)?;

    let (coeffs, commitment_points) =
        polynomial::generate_secret_polynomial(secret, coefficients);
    let commitment = VssCommitment {
        coeffs: commitment_points,
    };
    let vk = commitment.to_verifying_key();
    let vk_bytes = vk.serialize();
    let identifier = Identifier::derive(&vk_bytes)?;

    let sig = compute_proof_of_knowledge(&identifier, &coeffs, &vk, rng)?;

    let secret_pkg = Round1SecretPackage {
        identifier: identifier.clone(),
        coefficients: coeffs,
        commitment: commitment.clone(),
        min_signers,
        max_signers,
    };
    let pub_pkg = Round1Package {
        commitment,
        proof_of_knowledge: sig,
        verifying_key: vk,
    };
    Ok((secret_pkg, pub_pkg))
}

// ---------------------------------------------------------------------------
// DKG Part 2 (matches Dart dkgPart2 in dkg.dart:336)
// ---------------------------------------------------------------------------

/// DKG round 2: verify others' round 1 packages and compute shares for each.
///
/// - `secret_pkg`: our round 1 secret package.
/// - `round1_pkgs`: round 1 packages from all other participants (keyed by their identifier).
/// - Returns `(Round2SecretPackage, Map<Identifier, Round2Package>)`.
pub fn dkg_part2(
    secret_pkg: &Round1SecretPackage,
    round1_pkgs: &BTreeMap<Identifier, Round1Package>,
) -> Result<(Round2SecretPackage, BTreeMap<Identifier, Round2Package>), Error> {
    if round1_pkgs.len() != secret_pkg.max_signers - 1 {
        return Err(Error::IncorrectNumberOfPackages);
    }
    for pkg in round1_pkgs.values() {
        if pkg.commitment.coeffs.len() != secret_pkg.min_signers {
            return Err(Error::IncorrectNumberOfCommitments);
        }
    }

    let mut out = BTreeMap::new();
    for (sender_id, pkg) in round1_pkgs {
        let vk = pkg.commitment.to_verifying_key();
        verify_proof_of_knowledge(sender_id, &vk, &pkg.proof_of_knowledge)?;

        let share = polynomial::evaluate_polynomial(sender_id, &secret_pkg.coefficients);
        out.insert(sender_id.clone(), Round2Package { secret_share: share });
    }

    let fii = polynomial::evaluate_polynomial(
        &secret_pkg.identifier,
        &secret_pkg.coefficients,
    );

    Ok((
        Round2SecretPackage {
            identifier: secret_pkg.identifier.clone(),
            commitment: secret_pkg.commitment.clone(),
            secret_share: fii,
            min_signers: secret_pkg.min_signers,
            max_signers: secret_pkg.max_signers,
        },
        out,
    ))
}

// ---------------------------------------------------------------------------
// DKG Part 3 (matches Dart dkgPart3 in dkg.dart:377)
// ---------------------------------------------------------------------------

/// DKG round 3: verify received shares and compute final key package.
///
/// - `_r1_secret`: round 1 secret package (kept for API compatibility with Dart).
/// - `r2_secret`: round 2 secret package.
/// - `round1_pkgs`: others' round 1 packages.
/// - `round2_pkgs`: others' round 2 packages (shares addressed to us).
/// - Returns `(KeyPackage, PublicKeyPackage)` both normalized to even Y.
pub fn dkg_part3(
    _r1_secret: &Round1SecretPackage,
    r2_secret: &Round2SecretPackage,
    round1_pkgs: &BTreeMap<Identifier, Round1Package>,
    round2_pkgs: &BTreeMap<Identifier, Round2Package>,
) -> Result<(crate::keys::KeyPackage, PublicKeyPackage), Error> {
    if round1_pkgs.len() != r2_secret.max_signers - 1 {
        return Err(Error::IncorrectNumberOfPackages);
    }
    if round1_pkgs.len() != round2_pkgs.len() {
        return Err(Error::IncorrectNumberOfPackages);
    }
    for id in round1_pkgs.keys() {
        if !round2_pkgs.contains_key(id) {
            return Err(Error::IncorrectPackageMapping);
        }
    }

    let mut si = Scalar::ZERO;

    for (sender_id, pkg2) in round2_pkgs {
        let r1 = round1_pkgs
            .get(sender_id)
            .ok_or(Error::UnknownIdentifier)?;

        // Verify: share * G == commitment.getVerifyingShare(our_id)
        let share_point = point::base_mul(&pkg2.secret_share);
        let expected = r1.commitment.get_verifying_share(&r2_secret.identifier);
        if !point::points_equal(&share_point, &expected) {
            return Err(Error::InvalidSecretShare);
        }

        si = si + pkg2.secret_share;
    }

    // Add our own self-share
    si = si + r2_secret.secret_share;
    let secret_share = si;
    let verifying_share = point::base_mul(&secret_share);

    // Build commitment map for PublicKeyPackage
    let mut commit_map: BTreeMap<Identifier, VssCommitment> = BTreeMap::new();
    for (id, pkg) in round1_pkgs {
        commit_map.insert(id.clone(), pkg.commitment.clone());
    }
    commit_map.insert(r2_secret.identifier.clone(), r2_secret.commitment.clone());

    let public_key_package = pkp_from_dkg_commitments(&commit_map)?;

    let key_package = crate::keys::KeyPackage {
        identifier: r2_secret.identifier.clone(),
        secret_share,
        verifying_share,
        verifying_key: public_key_package.verifying_key.clone(),
        min_signers: r2_secret.min_signers,
    };

    Ok((key_package.into_even_y(), public_key_package.into_even_y()))
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

/// Build a PublicKeyPackage from the combined DKG commitments.
fn pkp_from_dkg_commitments(
    commits: &BTreeMap<Identifier, VssCommitment>,
) -> Result<PublicKeyPackage, Error> {
    let ids: Vec<Identifier> = commits.keys().cloned().collect();
    let list: Vec<VssCommitment> = commits.values().cloned().collect();
    let group = vss::sum_commitments(&list)?;
    Ok(pkp_from_commitment(&ids, &group))
}

/// Build a PublicKeyPackage from a summed commitment and the set of identifiers.
fn pkp_from_commitment(
    ids: &[Identifier],
    commit: &VssCommitment,
) -> PublicKeyPackage {
    let mut vmap = BTreeMap::new();
    for id in ids {
        vmap.insert(id.clone(), commit.get_verifying_share(id));
    }
    let vk = commit.to_verifying_key();
    PublicKeyPackage {
        verifying_shares: vmap,
        verifying_key: vk,
    }
}

// ---------------------------------------------------------------------------
// JSON serialization (wire-compatible with Dart)
// ---------------------------------------------------------------------------

impl DkgSignature {
    /// Deserialize from JSON: {"R": "hex_compressed", "Z": "hex_scalar"}
    pub fn from_json_value(v: &serde_json::Value) -> Result<Self, Error> {
        let r_hex = v["R"].as_str().ok_or(Error::SerializationError)?;
        let z_hex = v["Z"].as_str().ok_or(Error::SerializationError)?;

        let r_bytes = hex_decode_33(r_hex)?;
        let z_bytes = hex_decode_32(z_hex)?;

        let r = point::deserialize_compressed(&r_bytes)?;
        let z = scalar_from_bytes(&z_bytes)?;

        Ok(Self { r, z })
    }

    /// Serialize to JSON: {"R": "hex_compressed", "Z": "hex_scalar"}
    pub fn to_json_value(&self) -> serde_json::Value {
        let r_hex = hex_encode(&point::serialize_compressed(&self.r));
        let z_hex = hex_encode(&scalar_to_bytes(&self.z));
        serde_json::json!({
            "R": r_hex,
            "Z": z_hex
        })
    }
}

impl Round1Package {
    /// Deserialize from JSON (matching Dart Round1Package.fromJson).
    ///
    /// ```json
    /// {
    ///   "commitment": ["hex_point", ...],
    ///   "proofOfKnowledge": {"R": "hex", "Z": "hex"},
    ///   "verifyingKey": {"E": [byte, byte, ...]}
    /// }
    /// ```
    pub fn from_json(json: &str) -> Result<Self, Error> {
        let v: serde_json::Value =
            serde_json::from_str(json).map_err(|_| Error::SerializationError)?;
        Self::from_json_value(&v)
    }

    pub fn from_json_value(v: &serde_json::Value) -> Result<Self, Error> {
        let commitment = VssCommitment::from_json_value(&v["commitment"])?;
        let proof_of_knowledge = DkgSignature::from_json_value(&v["proofOfKnowledge"])?;

        // VerifyingKey: {"E": [byte, byte, ...]} — array of integers (33 compressed bytes)
        let e_arr = v["verifyingKey"]["E"]
            .as_array()
            .ok_or(Error::SerializationError)?;
        let mut e_bytes = [0u8; 33];
        if e_arr.len() != 33 {
            return Err(Error::SerializationError);
        }
        for (i, val) in e_arr.iter().enumerate() {
            e_bytes[i] = val.as_u64().ok_or(Error::SerializationError)? as u8;
        }
        let verifying_key = VerifyingKey::deserialize(&e_bytes)?;

        Ok(Self {
            commitment,
            proof_of_knowledge,
            verifying_key,
        })
    }

    /// Serialize to JSON (matching Dart Round1Package.toJson).
    pub fn to_json(&self) -> String {
        let v = self.to_json_value();
        serde_json::to_string(&v).unwrap_or_default()
    }

    pub fn to_json_value(&self) -> serde_json::Value {
        let vk_compressed = self.verifying_key.serialize();
        let e_arr: Vec<serde_json::Value> = vk_compressed
            .iter()
            .map(|&b| serde_json::Value::Number(serde_json::Number::from(b)))
            .collect();

        serde_json::json!({
            "commitment": self.commitment.to_json_value(),
            "proofOfKnowledge": self.proof_of_knowledge.to_json_value(),
            "verifyingKey": { "E": e_arr }
        })
    }
}

impl Round2Package {
    /// Deserialize from JSON: {"secretShare": "hex_scalar"}
    pub fn from_json(json: &str) -> Result<Self, Error> {
        let v: serde_json::Value =
            serde_json::from_str(json).map_err(|_| Error::SerializationError)?;
        Self::from_json_value(&v)
    }

    pub fn from_json_value(v: &serde_json::Value) -> Result<Self, Error> {
        let hex_str = v["secretShare"]
            .as_str()
            .ok_or(Error::SerializationError)?;
        let bytes = hex_decode_32(hex_str)?;
        let secret_share = scalar_from_bytes(&bytes)?;
        Ok(Self { secret_share })
    }

    /// Serialize to JSON: {"secretShare": "hex_scalar"}
    pub fn to_json(&self) -> String {
        let v = self.to_json_value();
        serde_json::to_string(&v).unwrap_or_default()
    }

    pub fn to_json_value(&self) -> serde_json::Value {
        let hex_str = hex_encode(&scalar_to_bytes(&self.secret_share));
        serde_json::json!({ "secretShare": hex_str })
    }
}

// ---------------------------------------------------------------------------
// Hex helpers
// ---------------------------------------------------------------------------

fn hex_encode(bytes: &[u8]) -> String {
    use alloc::format;
    bytes.iter().map(|b| format!("{:02x}", b)).collect()
}

fn hex_decode_32(s: &str) -> Result<[u8; 32], Error> {
    let bytes = hex_decode(s)?;
    if bytes.len() != 32 {
        return Err(Error::SerializationError);
    }
    let mut out = [0u8; 32];
    out.copy_from_slice(&bytes);
    Ok(out)
}

fn hex_decode_33(s: &str) -> Result<[u8; 33], Error> {
    let bytes = hex_decode(s)?;
    if bytes.len() != 33 {
        return Err(Error::SerializationError);
    }
    let mut out = [0u8; 33];
    out.copy_from_slice(&bytes);
    Ok(out)
}

fn hex_decode(s: &str) -> Result<Vec<u8>, Error> {
    if s.len() % 2 != 0 {
        return Err(Error::SerializationError);
    }
    let mut out = Vec::with_capacity(s.len() / 2);
    for i in (0..s.len()).step_by(2) {
        let byte = u8::from_str_radix(&s[i..i + 2], 16)
            .map_err(|_| Error::SerializationError)?;
        out.push(byte);
    }
    Ok(out)
}
